File: src/main/java/com/example/demo/util/DbUtils.java
package com.example.demo.util;

import lombok.experimental.UtilityClass;
import lombok.extern.slf4j.Slf4j;

import java.util.Arrays;
import java.util.function.Supplier;

/**
 * DB call wrapper that logs failures and rethrows.
 */
@Slf4j
@UtilityClass
public class DbUtils {

    public <T> T dbCall(String action, Supplier<T> block, Object... context) {
        try {
            return block.get();
        } catch (Exception ex) {
            String ctx = Arrays.stream(context).map(String::valueOf).reduce((a, b) -> a + "," + b).orElse("");
            log.error("DB {} failed; context={}", action, ctx, ex);
            throw ex;
        }
    }
}

File: src/main/kotlin/com/example/demo/util/DbUtils.kt
package com.example.demo.util

import org.slf4j.LoggerFactory

/**
 * DB 呼び出しを共通化するユーティリティ。例外発生時にログを出して再送出する。
 * 使用例: dbCall("selectByUser", userName) { mapper.selectByUser(userName) }
 */
inline fun <T> dbCall(action: String, vararg context: Any?, block: () -> T): T {
    try {
        return block()
    } catch (ex: Exception) {
        val logger = LoggerFactory.getLogger("com.example.demo.util.DbUtils")
        logger.error("DB {} failed; context={}", action, context.joinToString(), ex)
        throw ex
    }
}

